### PR TITLE
fix: Core returns wrong http code if if non match condition fails #560

### PR DESCRIPTION
**Pull request by astsiapanay**

### CODE REVIEW

File: storage/src/main/java/com/epam/aidial/core/storage/util/EtagHeader.java [hunk 1]
```diff
@@ -38,21 +43,38 @@ public void validate(Supplier<String> etagSupplier) {
...
+        validateIfMatch(etag);
+
+        validateIfNoneMatch(etag);
+    }
+
+    private void validateIfNoneMatch(String etag) {
         // If-None-Match used with the * value can be used to save a file not known to exist,
         // guaranteeing that another upload didn't happen before
         if (ifNoneMatchTags == null) {
             throw new HttpException(HttpStatus.PRECONDITION_FAILED, "Resource already exists");
```
Comment (artsiomkorzun) [inline storage/src/main/java/com/epam/aidial/core/storage/util/EtagHeader.java]: Should throw 304 for GET/HEAD?

File: storage/src/main/java/com/epam/aidial/core/storage/http/HttpStatus.java [hunk 2]
```diff
@@ -8,6 +8,7 @@
 public enum HttpStatus {
 
     OK(200),
+    NOT_MODIFIED(304),
```
Comment (artsiomkorzun) [inline storage/src/main/java/com/epam/aidial/core/storage/http/HttpStatus.java:11]: update fromStatusCode

File: server/src/main/java/com/epam/aidial/core/server/ProxyContext.java [hunk 3]
```diff
@@ -157,10 +157,20 @@ public Future<?> respond(HttpStatus status, String body) {
 
     public Future<?> respond(Throwable error, String fallbackError) {
         return error instanceof HttpException exception
-                ? respond(exception.getStatus(), exception.getMessage())
+                ? respond(exception)
                 : respond(HttpStatus.INTERNAL_SERVER_ERROR, fallbackError);
     }
 
+    public Future<?> respond(HttpException exception) {
+        Map<String, String> headers = exception.getHeaders();
```
Comment (artsiomkorzun) [inline server/src/main/java/com/epam/aidial/core/server/ProxyContext.java]: nit: you can make it return non-nullable and use response.headers().putAll()

File: storage/src/main/java/com/epam/aidial/core/storage/util/EtagHeader.java (hunk 1 above)
Comment (astsiapanay) [reply storage/src/main/java/com/epam/aidial/core/storage/util/EtagHeader.java]: right. fixed

File: storage/src/main/java/com/epam/aidial/core/storage/http/HttpStatus.java (hunk 2 above)
Comment (astsiapanay) [reply storage/src/main/java/com/epam/aidial/core/storage/http/HttpStatus.java:11]: fixed

File: server/src/main/java/com/epam/aidial/core/server/ProxyContext.java (hunk 3 above)
Comment (astsiapanay) [reply server/src/main/java/com/epam/aidial/core/server/ProxyContext.java]: fixed

**Review by artsiomkorzun — APPROVED** [review]

**Comment by astsiapanay** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-561.nightly-test.deltixhub.io
>E2E tests status: failed

**Pull request merged**